File: bin/migrate_config.dart
/// CLI command: `dart run saropa_lints migrate-config`
///
/// Moves `log_level`, `lane`, `memory_mode`, and `rule_packs` from the legacy
/// `plugins > saropa_lints:` block in `analysis_options.yaml` to top-level
/// keys in `analysis_options_custom.yaml`, eliminating false
/// `unsupported_option` warnings from the Dart SDK's plugin-block validator.
///
/// Safe to run multiple times — already-migrated keys are skipped.
library;

// ignore_for_file: avoid_print

import 'dart:io';

import 'package:saropa_lints/src/config/analysis_options_rule_packs.dart'
    show parseRulePacksEnabledList;
import 'package:saropa_lints/src/config/runtime_tier_cap.dart'
    show parseScalarFromPluginBlock;
import 'package:saropa_lints/src/init/custom_overrides_core.dart'
    show writeRulePacksToCustomFile;

/// Scalar keys that were moved from the plugin block to the custom file.
const _migrateKeys = <String>{'log_level', 'lane', 'memory_mode'};

/// Pattern matching the indented `rule_packs:` block (with `enabled:` list)
/// inside the plugin block. Used to remove it from analysis_options.yaml.
final _pluginRulePacksBlock = RegExp(
  r'^[ \t]+rule_packs:\s*\n(?:[ \t]+enabled:\s*\n)?(?:[ \t]+-\s+\S+.*\n|[ \t]+#[^\n]*\n|[ \t]*\n)*',
  multiLine: true,
);

/// Entry point for `dart run saropa_lints migrate-config`.
///
/// Flags:
///   `--dry-run`  Preview what would change without writing any files.
///   First positional arg is the project directory (defaults to `.`).
void main(List<String> args) {
  // Parse --dry-run flag and extract the directory argument.
  final dryRun = args.contains('--dry-run');
  final positional = args.where((a) => a != '--dry-run').toList();
  final dir = positional.isNotEmpty ? positional.first : '.';
  final sep = Platform.pathSeparator;

  final mainFile = File('$dir${sep}analysis_options.yaml');
  if (!mainFile.existsSync()) {
    print('No analysis_options.yaml found in $dir');
    exitCode = 1;
    return;
  }

  // Normalize line endings so removal regexes (which match `\n`) work on
  // Windows files that may contain `\r\n`.
  final mainContent =
      mainFile.readAsStringSync().replaceAll('\r\n', '\n').replaceAll('\r', '\n');

  // --- Scalar keys (log_level, lane, memory_mode) ---
  final found = <String, String>{};
  for (final key in _migrateKeys) {
    final value = parseScalarFromPluginBlock(mainContent, {key});
    if (value != null) {
      found[key] = value;
    }
  }

  // --- Nested key: rule_packs ---
  final rulePackIds = parseRulePacksEnabledList(mainContent);

  if (found.isEmpty && rulePackIds.isEmpty) {
    print('Nothing to migrate — no legacy config keys found under '
        'plugins > saropa_lints:');
    return;
  }

  // Read or create the custom file.
  final customFile = File('$dir${sep}analysis_options_custom.yaml');
  var customContent = customFile.existsSync()
      ? customFile.readAsStringSync().replaceAll('\r\n', '\n').replaceAll('\r', '\n')
      : '';

  // Only add scalar keys that aren't already in the custom file, but always
  // remove them from the legacy plugin block (even when skipped) so the
  // unsupported_option warning is eliminated.
  final added = <String>[];
  for (final entry in found.entries) {
    // Check if already present as a top-level key.
    final existing = RegExp(
      '^${entry.key}:\\s',
      multiLine: true,
    ).hasMatch(customContent);
    if (existing) {
      print('  SKIP ${entry.key}: already in analysis_options_custom.yaml');
    } else {
      // Append the key to the custom file.
      if (customContent.isNotEmpty && !customContent.endsWith('\n')) {
        customContent += '\n';
      }
      customContent += '${entry.key}: ${entry.value}\n';
      print('  MOVE ${entry.key}: ${entry.value}');
    }
    // Track all found keys for removal from the main file, regardless of
    // whether they were added to or already existed in the custom file.
    added.add(entry.key);
  }

  // Write the custom file with the scalar keys first.
  if (added.isNotEmpty && !dryRun) {
    customFile.writeAsStringSync(customContent);
  }

  // Migrate rule_packs: write to custom file, remove from main file.
  // Always remove from the legacy block even when skipped, to eliminate
  // the unsupported_option warning.
  var hasLegacyRulePacks = rulePackIds.isNotEmpty;
  if (rulePackIds.isNotEmpty) {
    // Check if rule_packs already exists in the custom file — reuse the
    // in-memory content (may have been updated with scalar keys above).
    final customHasRulePacks = RegExp(
      r'^rule_packs:\s',
      multiLine: true,
    ).hasMatch(customContent);
    if (customHasRulePacks) {
      print('  SKIP rule_packs: already in analysis_options_custom.yaml');
    } else {
      // Write rule_packs block to the custom file using the shared helper.
      if (!dryRun) {
        writeRulePacksToCustomFile(customFile, rulePackIds);
      }
      print('  MOVE rule_packs: [${rulePackIds.join(', ')}]');
    }
  }

  if (found.isEmpty && !hasLegacyRulePacks) {
    print('All keys already migrated — nothing to do.');
    return;
  }

  // Remove the scalar keys from the main file's plugin block. Match a full
  // line like `    log_level: info  # comment` and remove it (including the
  // newline). Indented so it's inside the plugin block, not top-level.
  var updatedMain = mainContent;
  for (final key in added) {
    updatedMain = updatedMain.replaceAll(
      RegExp('^[ \\t]+$key:\\s+[^\\n]*\\n', multiLine: true),
      '',
    );
  }

  // Remove the rule_packs block from the main file (even when skipped —
  // the goal is to eliminate the warning, not just copy the value).
  if (hasLegacyRulePacks) {
    updatedMain = updatedMain.replaceAll(_pluginRulePacksBlock, '');
  }

  if (!dryRun) {
    mainFile.writeAsStringSync(updatedMain);
  }

  final totalMoved = added.length + (hasLegacyRulePacks ? 1 : 0);
  print('');
  if (dryRun) {
    print('[dry-run] Would migrate $totalMoved key(s) to '
        'analysis_options_custom.yaml. No files were modified.');
  } else {
    print('Migrated $totalMoved key(s) to analysis_options_custom.yaml.');
    print('The unsupported_option warnings will no longer appear.');
  }
}

File: extension/src/config/migrateConfig.ts
/**
 * Migrates `log_level`, `lane`, `memory_mode`, and `rule_packs` from the
 * legacy `plugins > saropa_lints:` block in `analysis_options.yaml` to
 * top-level keys in `analysis_options_custom.yaml`.
 *
 * Eliminates false `unsupported_option` warnings from the Dart SDK's
 * plugin-block validator, which hardcodes allowed keys.
 *
 * Surfaced in the sidebar (Config > Migrate config keys), the command palette
 * (`Saropa Lints: Migrate Config Keys`), and the CLI
 * (`dart run saropa_lints migrate-config`).
 */
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as vscode from 'vscode';
import { parseLaneFromPluginBlock } from './laneConfig';
import { parseRulePacksEnabled, removeLegacyRulePacksFromMainFile, writeRulePacksEnabled } from '../rulePacks/rulePackYaml';

/** Keys that were moved from the plugin block to the custom file. */
const MIGRATE_KEYS = ['log_level', 'lane', 'memory_mode'] as const;

/** Counts leading whitespace (spaces and tabs). */
function leadingWhitespace(value: string): number {
  let count = 0;
  while (count < value.length && (value[count] === ' ' || value[count] === '\t')) count++;
  return count;
}

/**
 * Reads a scalar value from `plugins > saropa_lints:` block.
 * Mirrors Dart's `parseScalarFromPluginBlock` for the specific migrate keys.
 */
function parseScalarFromPluginBlock(content: string, key: string): string | undefined {
  const lines = content.replace(/\r\n/g, '\n').replace(/\r/g, '\n').split('\n');
  for (let i = 0; i < lines.length; i++) {
    const trimmed = lines[i].replace(/\s+$/, '');
    if (!/^\s+saropa_lints:\s*(#.*)?$/.test(trimmed)) continue;
    const baseIndent = leadingWhitespace(lines[i]);
    for (let j = i + 1; j < lines.length; j++) {
      const inner = lines[j];
      const t = inner.replace(/^\s+/, '');
      if (t.length === 0 || t.startsWith('#')) continue;
      const ind = leadingWhitespace(inner);
      if (ind <= baseIndent) break;
      const re = new RegExp(`^\\s*${key}:\\s*([^\\s#]+)`);
      const m = re.exec(inner);
      if (m) {
        const v = m[1]?.trim();
        if (!v) continue;
        return v.replace(/^['"]|['"]$/g, '').toLowerCase();
      }
    }
  }
  return undefined;
}

/** Result of a migration attempt. */
export interface MigrateResult {
  /** Keys that were successfully moved. */
  moved: string[];
  /** Keys that were already in the custom file (skipped). */
  skipped: string[];
  /** True when --dry-run was used (no files were modified). */
  dryRun?: boolean;
  /** Error message if the operation failed entirely. */
  error?: string;
}

/**
 * Runs the migration: reads legacy keys from the plugin block, adds them
 * to the custom file, removes them from the main file.
 *
 * Pass `dryRun: true` to preview changes without modifying any files.
 */
export function migrateConfigKeys(root: string, { dryRun = false } = {}): MigrateResult {
  const mainPath = path.join(root, 'analysis_options.yaml');
  if (!fs.existsSync(mainPath)) {
    return { moved: [], skipped: [], error: 'analysis_options.yaml not found' };
  }

  const mainContent = fs.readFileSync(mainPath, 'utf8');
  const customPath = path.join(root, 'analysis_options_custom.yaml');

  // Find which keys exist in the old plugin block.
  const found = new Map<string, string>();
  for (const key of MIGRATE_KEYS) {
    const value = parseScalarFromPluginBlock(mainContent, key);
    if (value !== undefined) {
      found.set(key, value);
    }
  }

  if (found.size === 0) {
    return { moved: [], skipped: [] };
  }

  // Read or create the custom file.
  let customContent = '';
  try {
    customContent = fs.readFileSync(customPath, 'utf8');
  } catch {
    // File doesn't exist yet — will create.
  }

  const moved: string[] = [];
  const skipped: string[] = [];

  for (const [key, value] of found) {
    // Skip if already present as a top-level key in the custom file.
    if (new RegExp(`^${key}:\\s`, 'm').test(customContent)) {
      skipped.push(key);
      continue;
    }
    // Append the key.
    if (customContent.length > 0 && !customContent.endsWith('\n')) {
      customContent += '\n';
    }
    customContent += `${key}: ${value}\n`;
    moved.push(key);
  }

  // Write the custom file with the scalar keys (only if any were added).
  const scalarKeysChanged = moved.length > 0;
  if (scalarKeysChanged && !dryRun) {
    fs.writeFileSync(customPath, customContent, 'utf8');
  }

  // Remove ALL found scalar keys from the main file's plugin block — even
  // skipped ones — so the unsupported_option warning is eliminated.
  if (found.size > 0 && !dryRun) {
    let updatedMain = mainContent;
    for (const key of found.keys()) {
      updatedMain = updatedMain.replace(
        new RegExp(`^[ \\t]+${key}:\\s+[^\\n]*\\n`, 'm'),
        '',
      );
    }
    if (updatedMain !== mainContent) {
      fs.writeFileSync(mainPath, updatedMain, 'utf8');
    }
  }

  // Migrate rule_packs (nested block) from the plugin block to custom file.
  const rulePackIds = parseRulePacksEnabled(mainContent);
  if (rulePackIds.length > 0) {
    // Check if rule_packs already exists in the custom file.
    const latestCustom = fs.existsSync(customPath)
      ? fs.readFileSync(customPath, 'utf8')
      : '';
    if (/^rule_packs:\s/m.test(latestCustom)) {
      skipped.push('rule_packs');
    } else {
      // writeRulePacksEnabled writes to custom file and cleans up main file.
      if (!dryRun) {
        writeRulePacksEnabled(root, rulePackIds);
      }
      moved.push('rule_packs');
    }
    // Even if rule_packs was skipped (already in custom file), clean up the
    // legacy block from the main file to stop the warning.
    if (!moved.includes('rule_packs') && !dryRun) {
      // removeLegacyRulePacksFromMainFile is called by writeRulePacksEnabled
      // when it runs; only need manual cleanup for the skip path.
      removeLegacyRulePacksFromMainFile(root);
    }
  }

  return { moved, skipped, dryRun: dryRun || undefined };
}

/**
 * VS Code command handler for `saropaLints.migrateConfig`.
 * Shows a progress toast and reports results.
 */
export async function runMigrateConfig(root: string | undefined): Promise<void> {
  if (!root) {
    void vscode.window.showWarningMessage(
      'No Dart project root found. Open a project with analysis_options.yaml first.',
    );
    return;
  }

  const result = migrateConfigKeys(root);

  if (result.error) {
    void vscode.window.showErrorMessage(`Config migration failed: ${result.error}`);
    return;
  }

  if (result.moved.length === 0 && result.skipped.length === 0) {
    void vscode.window.showInformationMessage(
      'Nothing to migrate — no legacy config keys found under plugins > saropa_lints:',
    );
    return;
  }

  if (result.moved.length === 0) {
    void vscode.window.showInformationMessage(
      `All keys already migrated (${result.skipped.join(', ')}).`,
    );
    return;
  }

  void vscode.window.showInformationMessage(
    `Migrated ${result.moved.join(', ')} to analysis_options_custom.yaml. ` +
    `The unsupported_option warnings will no longer appear.`,
  );
}
